File: leetcode/typescript/000383.ransom-note.ts
function canConstruct(ransomNote: string, magazine: string): boolean {
  const magazineLetterMap = new Map<string, number>();

  for (const letter of magazine) {
    const letterCount = magazineLetterMap.get(letter);

    if (letterCount) {
      magazineLetterMap.set(letter, letterCount + 1);
    } else {
      magazineLetterMap.set(letter, 1);
    }
  }

  for (const letter of ransomNote) {
    const letterCountLeftInMagazine = magazineLetterMap.get(letter);
    if (!letterCountLeftInMagazine) return false;

    magazineLetterMap.set(letter, letterCountLeftInMagazine - 1);
  }

  return true;
}
// Time complexity: O(n)
// Space complexity: O(n)

File: leetcode/python/000383.ransom-note.py
from collections import Counter


class Solution:
    # Time complexity: O(n)
    # Space complexity: O(1) - max length 26
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        magazine_counter = Counter(magazine)
        for ch in ransomNote:
            if magazine_counter[ch] == 0:
                return False
            magazine_counter[ch] -= 1
        return True
